hotshot-state-prover/src/service.rs

```
@@ -0,0 +1,575 @@
+
+/// submit the latest finalized state along with a proof to the L1 LightClient contract
+pub async fn submit_state_and_proof(
+    proof: Proof,
+    public_input: PublicInput,
+    config: &StateProverConfig,
+) -> Result<(), ProverError> {
+    let contract = prepare_contract(config).await?;
+
+    // prepare the input the contract call and the tx itself
+    let proof: ParsedPlonkProof = proof.into();
+    let new_state: ParsedLightClientState = public_input.into();
+    let tx = contract.new_finalized_state(new_state.into(), proof.into());
+
+    // send the tx
+    let (_receipt, _included_block) = sequencer_utils::contract_send(&tx)
+        .await
+        .map_err(ProverError::ContractError)?;
+
+    Ok(())
+}
+
+pub async fn sync_state(
+    st: &StakeTable<BLSPubKey, StateVerKey, CircuitField>,
+    proving_key: &ProvingKey,
+    relay_server_client: &Client<ServerError>,
+    config: &StateProverConfig,
+) -> Result<(), ProverError> {
+    tracing::info!("Start syncing light client state.");
+
+    let bundle = fetch_latest_state(relay_server_client).await?;
+    let old_state = read_contract_state(config).await?;
+    if old_state.block_height >= bundle.state.block_height {
+        tracing::info!("No update needed.");
+        return Ok(());
+    }
+
+    let threshold = st.total_stake(SnapshotVersion::LastEpochStart)? * 2 / 3;
+    let entries = st
+        .try_iter(SnapshotVersion::LastEpochStart)
+        .unwrap()
+        .map(|(_, stake_amount, state_key)| (state_key, stake_amount))
+        .collect::<Vec<_>>();
+    let mut signer_bit_vec = vec![false; entries.len()];
+    let mut signatures = vec![Default::default(); entries.len()];
+    let mut accumulated_weight = U256::zero();
+    entries.iter().enumerate().for_each(|(i, (key, stake))| {
+        if let Some(sig) = bundle.signatures.get(key) {
+            signer_bit_vec[i] = true;
+            signatures[i] = sig.clone();
+            accumulated_weight += *stake;
+        }
+    });
+
```

Maybe throw an error if `accumulated_weight < threshold` instead of getting the error while generating the proof.

---

Done in [f005900](https://github.com/EspressoSystems/espresso-sequencer/pull/1041/commits/f0059003d42ca79309513fe94c1ffc647d3e474b)